### PR TITLE
Fixed a transcoding bug that occurred when remote transcoder was removed

### PR DESCRIPTION
**Pull request by eliteprox**
**What does this pull request do? Explain your changes. (required)**
This change fixes a bug that caused an orchestrator to lose all transcoders when one of it's transcoders disconnects unexpectedly. 

**Specific updates (required)**

- Simplified logic in `removeFromRemoteTranscoders` so that correct transcoder is removed from RemoteTranscoderManager

**How did you test each of these updates (required)**
We tested up to 4 streams going to an orchestrator with up to 4 remote transcoders. We stopped one of the transcoders in the following scenarios and observed streams move to another transcoder without disconnecting any other transcoders.
- 1 stream on each of the 4 remote transcoders
- 2 streams on each of 2 remote transcoders
- 4 streams on 1 transcoder with 3 additional transcoders having 0 load. 

In every case we observed the streams safely move to another available transcoder without causing other transcoders to break.

**Does this pull request close any open issues?**

#2605
#2706

**Checklist:**

- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

**Comment by Titan-Node** [comment]
Works great, did many tests with removing and adding multiple Ts without any dropped streams.

**Comment by thomshutt** [comment]
This looks like a great simplification and as far as I can tell, the original code is just convoluted because of micro-optimisations that we don't need. Could you also please add a unit test so that we can avoid breaking this again in the future?

File: CHANGELOG_PENDING.md
Comment (leszko) [inline CHANGELOG_PENDING.md:25]: Why is this line removed?

**Comment by eliteprox** [comment]
> This looks like a great simplification and as far as I can tell, the original code is just convoluted because of micro-optimisations that we don't need. Could you also please add a unit test so that we can avoid breaking this again in the future?

Yep! I will write some tests, was thinking the same. This is a critical functionality. Thanks for looking it over 😎

**Review by leszko — COMMENTED** [review]
Nice, great simplification.

+1 for adding unit tests.

Also wondering, what's the functional difference between the old and the new code? Isn't it working exactly the same, how does it fix the issue you encountered?

File: CHANGELOG_PENDING.md
Comment (eliteprox) [reply CHANGELOG_PENDING.md:25]: This was a note from prior release that somehow was left in the pending changelog.

**Comment by stronk-dev** [comment]
Already has a unit test BTW:
https://github.com/livepeer/go-livepeer/blob/ab553b4c58589e1ce3bb41905dd1f56f804c3900/core/orch_test.go#L372-L420

**Comment by thomshutt** [comment]
Good point @stronk-dev - I'd still want to update that test such that it would've caught the case that was breaking though

**Comment by eliteprox** [comment]
> Also wondering, what's the functional difference between the old and the new code? Isn't it working exactly the same, how does it fix the issue you encountered?

It appears some funky logic or race condition led to it. I couldn't follow the logic there. I discovered this after monitoring my own node. We then tested with real test streams until it was a confirmed bug, and then worked through the solution. The issue occurred only when there was 2 or more streams on the transcoder that was disconnected.

I wrote a new test which reproduced the issue on the old version of `removeRemoteTranscoder` and passed on the new code.
https://github.com/livepeer/go-livepeer/blob/707115e481c487d806cba93769c6b3fe8b73a92f/core/orch_test.go#L423-L477

In the test which uses 2 remote transcoders and 8 streams:
Old code = Only 5 streams made it to the 2nd transcoder
New code = All 8 streams went to the 2nd transcoder

**Comment by eliteprox** [comment]
Added new test `TestRemoveFromRemoteTranscoders_StreamFailover`. I left the previous test because the new one is more focused on the high-level result of shifting streams to the opposite transcoder and uses the `selectTranscoder` function to validate this.

**Comment by eliteprox** [comment]
@leszko @thomshutt Any chance this can make it into 0.5.38? #2753

File: core/orch_test.go
Comment (leszko) [inline core/orch_test.go]: Won't this test be flaky? I think depending on timing (especially in ms) can make this test sometime pass, sometimes not. I'd avoid having time-dependent tests.

Comment (eliteprox) [reply core/orch_test.go]: I copied the wait commands from another test.  I think they are not needed. I didn't see anything in my development that indicates they are required, so I can try removing them.

Comment (leszko) [reply core/orch_test.go]: Yeah, then I suggest removing them if not needed. Then, Do you need to run these functions above in separate goroutines?

Comment (eliteprox) [reply core/orch_test.go]: Yes, apparently the separate go routine is required in this case, if not done with the go routine, the test will timeout. I tried removing the time.sleep between them, but it caused test to move on without transcoder connected at all.

I pushed a commit to cleanup a few related lines.

Comment (leszko) [reply core/orch_test.go]: Hmm, THB, in my opinion, it's better to not have tested with `time.Sleep()`, because they will sometimes fail sometimes pass in our CI. So, if it's not possible (or easy) to write a test without sleep, then I suggest removing this test and just merging without it. @thomshutt wdyt?

The improvement to the code is already good, so should be safe to merge.

**Comment by thomshutt** [comment]
@leszko @eliteprox I don't understand - if the problem is this method being called by multiple goroutines in an unsafe way then shouldn't the problem be fixed further up the stack? 

If the problem is inside `removeFromRemoteTranscoders` then we should be able to write a test that just calls that method, without any concurrency code.

**Comment by leszko** [comment]
> @leszko @eliteprox I don't understand - if the problem is this method being called by multiple goroutines in an unsafe way then shouldn't the problem be fixed further up the stack?
> 
> If the problem is inside `removeFromRemoteTranscoders` then we should be able to write a test that just calls that method, without any concurrency code.

From my current understanding, there are 2 issues:
- this function is unnecessarily overcomplicated => this PR solves this issue
- this function is used in an unsafe manner by multiple goroutines => this PR does not solve this issue (though might make it less visible)

Then, I suggest doing the following:
1. For this PR, write a very simple sequential unit test just to test the function `removeFromRemoteTranscoders()`, nothing more
2. Do not include the test `TestRemoveFromRemoteTranscoders_StreamFailover`, since I believe it's flaky
3. Merge this PR
4. Create a GH Issue to solve the concurrency issue (unless it's simple, then @eliteprox feel free to fix it in this PR)

**Comment by eliteprox** [comment]
> * For this PR, write a very simple sequential unit test just to test the function `removeFromRemoteTranscoders()`, nothing more
> 
> * Do not include the test `TestRemoveFromRemoteTranscoders_StreamFailover`, since I believe it's flaky
> 
> * Merge this PR
> 
> * Create a GH Issue to solve the concurrency issue (unless it's simple, then @eliteprox feel free to fix it in this PR)

Can we simply continue using the existing `TestRemoveFromRemoteTranscoders` test for now until the tests can be refactored, and merge this PR? This test properly covers removing transcoders within it's own scope, imho.

I've removed `TestRemoveFromRemoteTranscoders_StreamFailover`. I created GH issue #2756 for the refactoring of tests.

**Comment by leszko** [comment]
> > * For this PR, write a very simple sequential unit test just to test the function `removeFromRemoteTranscoders()`, nothing more
> > * Do not include the test `TestRemoveFromRemoteTranscoders_StreamFailover`, since I believe it's flaky
> > * Merge this PR
> > * Create a GH Issue to solve the concurrency issue (unless it's simple, then @eliteprox feel free to fix it in this PR)
> 
> Can we simply continue using the existing `TestRemoveFromRemoteTranscoders` test for now until the tests can be refactored, and merge this PR? This test properly covers removing transcoders within it's own scope, imho.
> 
> I've removed `TestRemoveFromRemoteTranscoders_StreamFailover`. I created GH issue #2756 for the refactoring of tests.

I'm ok with this. I think this change is safe as it is and it simplifies the code a lot. @thomshutt wdyt?

**Review by leszko — APPROVED** [review]
LGTM

Let's wait for @thomshutt to comment/approve and then we can merge.

**Comment by Titan-Node** [comment]
Been running this upgrade on prod for 4 days with no errors to report

**Review by thomshutt — APPROVED** [review]

**Comment by thomshutt** [comment]
LGTM, thanks for the fix and for responding to all our questions @eliteprox!

**Pull request merged**
